src/nodes/NodeConnectionManager.ts

```diff
@@ -0,0 +1,714 @@
+import type KeyManager from '../keys/KeyManager';
+import type ReverseProxy from '../network/ReverseProxy';
+import type ForwardProxy from '../network/ForwardProxy';
+import type { Host, Hostname, Port } from '../network/types';
+import type { ResourceAcquire } from '../utils';
+import type { Timer } from '../types';
+import type NodeGraph from './NodeGraph';
+import type { NodeId, NodeAddress, NodeData, SeedNodes } from './types';
+import Logger from '@matrixai/logger';
+import { StartStop, ready } from '@matrixai/async-init/dist/StartStop';
+import { IdInternal } from '@matrixai/id';
+import { status } from '@matrixai/async-init';
+import NodeConnection from './NodeConnection';
+import * as nodesUtils from './utils';
+import * as nodesErrors from './errors';
+import GRPCClientAgent from '../agent/GRPCClientAgent';
+import * as validationUtils from '../validation/utils';
+import * as networkUtils from '../network/utils';
+import * as agentErrors from '../agent/errors';
+import * as grpcErrors from '../grpc/errors';
+import * as nodesPB from '../proto/js/polykey/v1/nodes/nodes_pb';
+import { RWLock, withF } from '../utils';
+
+type ConnectionAndLock = {
+  connection?: NodeConnection<GRPCClientAgent>;
+  timer?: NodeJS.Timer;
+  lock: RWLock;
+};
+
+/**
+ * Data structure to store all NodeConnections. If a connection to a node n does
+ * not exist, no entry for n will exist in the map. Alternatively, if a
+ * connection is currently being instantiated by some thread, an entry will
+ * exist in the map, but only with the lock (no connection object). Once a
+ * connection is instantiated, the entry in the map is updated to include the
+ * connection object.
+ */
+type NodeConnectionMap = Map<string, ConnectionAndLock>;
```

Comment:
Because this is a `Map`, it does not need its key to be a string.
Instead you can use `Map<NodeId, ConnectionAndLock>`.

Then in all subsequent operations, you don't need to convert to and back to binary string.

Maps can use objects as keys, and it uses object identity. Which is exactly what we can do here.

Reply:
Ok so apparently we need to compare on string equality due to functions like `createConnection`. So to make this clearer, it should be `Map<NodeIdString, ConnectionAndLock>;`.

Reply:
This uses NodeIdString now. as mentioned using NodeId for the key in the map leads to issues where we can fail to .get(nodeId) even when we have the same nodeId. this comes up in testing a lot and possibly general use.